core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/batchutilities/BatchBufferTest.java

```diff
@@ -0,0 +1,269 @@
+        for (int i = 0; i < numThreads; i++) {
+            try {
+                CompletableFuture.allOf(completionService.take()
+                                                         .get(300, TimeUnit.MILLISECONDS)
+                                                         .toArray(new CompletableFuture[0]))
+                                 .join();
+            } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                System.err.println(e);
+            }
+        }
+
+        Iterator<CompletableFuture<String>> responsesIterator = responses.iterator();
+        for (int i = 0; responsesIterator.hasNext(); i++) {
+            System.err.println(responsesIterator.next().join());
+        }
+        Assert.assertEquals(responses.size(), numThreads*numMessages);
+    }
+
+    private void createThreadsAndSendMessages(int numThreads, int numMessages, String destination,
+                                              ConcurrentLinkedQueue<CompletableFuture<String>> responses,
+                                              CompletionService<List<CompletableFuture<String>>> completionService) {
+        String newDestination;
+        for (int i = 0; i < numThreads; i++) {
+            if (destination == null) {
+                newDestination = Integer.toString(i);
+            } else {
+                newDestination = destination;
+            }
+            sendRequestToDestination(i*numMessages, numMessages, newDestination, responses, completionService);
+        }
+    }
+
+    private void sendRequestToDestination(int startingId, int numMessages, String destination,
+                                          ConcurrentLinkedQueue<CompletableFuture<String>> responses,
+                                          CompletionService<List<CompletableFuture<String>>> completionService) {
+        String[] requests = createRequestsOfSize(startingId, numMessages);
+        completionService.submit(() -> {
+            List<CompletableFuture<String>> newResponses = createAndSendResponses(requests, destination);
+            responses.addAll(newResponses);
+            return newResponses;
+        });
+    }
+
+    private boolean waitForTime(int msToWait) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        try {
+            return countDownLatch.await(msToWait, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        return false;
+    }
+
+    private void checkAllResponses(List<CompletableFuture<String>> responses) {
+        for (int i = 0; i < responses.size(); i++) {
+            Assert.assertEquals(responses.get(i).join(), "Message " + i);
+        }
+    }
+
+    private String[] createRequestsOfSize(int size) {
+        return createRequestsOfSize(0, size);
+    }
+
+    private String[] createRequestsOfSize(int startingId, int size) {
```

Comment:
nit, is there any reason we use array and not List? List is a lot easier to work with

Reply:
Yeah I think I will change all tests to use lists. Definitely makes it a lot easier.